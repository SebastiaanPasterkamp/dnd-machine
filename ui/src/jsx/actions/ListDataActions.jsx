import React from 'react';
import Reflux from 'reflux';
import _ from 'lodash';

import ReportingActions from './ReportingActions.jsx';

function jsonOrBust(response) {
    if (response.ok) {
        return response.json();
    }
    throw response.statusText;
};

var ListDataActions = Reflux.createActions({
    "setState": {},
    "doLogin": {children: ['completed', 'failed']},
    "doLogout": {children: ['completed', 'failed']},
    "fetchItems": {children: ['completed', 'failed']},
});
let throttledGet = {};

ListDataActions.fetchItems.listen((type, category=null) => {
    let path = '/' + _.filter([
        category, type, category ? 'api' : null]).join('/');

    if (!(path in throttledGet)) {
        throttledGet[path] = _.throttle((path, type) => {

            fetch(path, {
                credentials: 'same-origin',
                'headers': {
                    'X-Requested-With': 'XMLHttpRequest'
                }
            })
            .then(jsonOrBust)
            .then((response) => {
                ListDataActions.fetchItems.completed({
                    [type]: response
                });
            })
            .catch((error) => {
                console.log(error);
                ListDataActions.fetchItems.failed(type, error);
            });
        }, 1000, {leading: true, trailing: false});
    }

    throttledGet[path](path, type);
});

ListDataActions.doLogin.listen((credentials, success, failure) => {
    fetch('/login', {
        credentials: 'same-origin',
        method: 'POST',
        'headers': {
            'X-Requested-With': 'XMLHttpRequest',
            'Content-Type': 'application/json',
        },
        body: JSON.stringify(credentials)
    })
    .then(jsonOrBust)
    .then((response) => {
        ListDataActions.doLogin.completed(response);
        ListDataActions.fetchItems.completed({
            current_user: response
        });

        ReportingActions.showMessage(
            'good',
            'Login successful',
            'Login',
            5
        );

        if (success) {
            success(content);
        }
    })
    .catch((error) => {
        console.log(error);

        ReportingActions.showMessage(
            'bad',
            'Login failed',
            'Login',
            10
        );

        ListDataActions.doLogin.failed(error);

        if (failure) {
            failure();
        }
    });
});

ListDataActions.doLogout.listen((callback) => {
    fetch('/logout', {
        credentials: 'same-origin',
        'headers': {
            'X-Requested-With': 'XMLHttpRequest'
        }
    })
    .then(jsonOrBust)
    .then((response) => {
        ListDataActions.doLogout.completed(response);
        ListDataActions.fetchItems.completed({
            current_user: null
        });

        ReportingActions.showMessage(
            'good',
            'Logout successful',
            'Logout',
            5
        );

        if (callback) {
            callback();
        }
    })
    .catch((error) => {
        console.log(error);
        ListDataActions.doLogout.failed(error);
        ListDataActions.fetchItems.completed({
            current_user: null
        });

        ReportingActions.showMessage(
            'bad',
            'Logout failed',
            'Logout',
            10
        );

        if (callback) {
            callback();
        }
    });
});

export default ListDataActions;
