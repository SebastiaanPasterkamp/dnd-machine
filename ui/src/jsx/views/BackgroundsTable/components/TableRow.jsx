import React from 'react';
import PropTypes from 'prop-types';
import MDReactComponent from 'markdown-react-js';

import BackgroundLinks from '../../../components/BackgroundLinks';

export const TableRow = function({ id, name, description }) {
    return (
        <tr data-name={id}>
            <th>
                { name }
                <BackgroundLinks
                    altStyle={true}
                    id={id}
                />
            </th>
            <td>
                <MDReactComponent
                    text={description || ''}
                />
            </td>
        </tr>
    );
};

TableRow.propTypes = {
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
    description: PropTypes.string,
};

TableRow.defaultProps = {
    description: '',
};

export default TableRow;
