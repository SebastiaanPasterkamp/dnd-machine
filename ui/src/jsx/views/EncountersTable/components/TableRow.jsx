import React from 'react';
import PropTypes from 'prop-types';
import {
    map,
} from 'lodash/fp';

import utils from '../../../utils';

import ListDataWrapper from '../../../hocs/ListDataWrapper';

import CampaignLabel from '../../../components/CampaignLabel';
import ChallengeRating from '../../../components/ChallengeRating';
import EncounterLinks from '../../../components/EncounterLinks';
import MonsterLabel from '../../../components/MonsterLabel';
import XpRating from '../../../components/XpRating';

const TableRow = function({
    id, campaign_id, hostedParty, name, monster_ids, challenge_rating,
    xp_rating, challenge_modified, xp, xp_modified,
}) {
    const ratingStyle = hostedParty
        ? utils.closest(
            {
                muted: 0,
                info: hostedParty.challenge.easy,
                good: hostedParty.challenge.medium,
                warning: hostedParty.challenge.hard,
                bad: hostedParty.challenge.deadly,
            },
            xp_rating,
            ratingStyle
        )
        : 'muted';

    return (
        <tr data-name={id}>
            <th>
                {name}
                <EncounterLinks
                    id={id}
                    altStyle={true}
                />
            </th>
            <td>
                {map(monster => (
                    <span key={monster.id}>
                        {monster.count}
                        &nbsp;x&nbsp;
                        <MonsterLabel
                            monster_id={monster.id}
                        />
                    </span>
                ))(monster_ids)}
            </td>
            <td className={ratingStyle}>
                <span>
                    <strong>Challenge:</strong>&nbsp;
                    <ChallengeRating challengeRating={challenge_rating} />
                    &nbsp;/&nbsp;
                    <XpRating xpRating={xp} />
                </span>
                {hostedParty ? (
                    <span>
                        <strong>Modified:</strong>&nbsp;
                        <ChallengeRating challengeRating={challenge_modified} />
                        &nbsp;/&nbsp;
                        <XpRating xpRating={xp_modified} />
                    </span>
                ) : null}
            </td>
            <td>
                {campaign_id ? (
                    <CampaignLabel id={campaign_id} />
                ) : null }
            </td>
        </tr>
    );
};

TableRow.propTypes = {
    id: PropTypes.number.isRequired,
    campaign_id: PropTypes.number,
    hostedParty: PropTypes.shape({
        challenge: PropTypes.shape({
            easy: PropTypes.number,
            medium: PropTypes.number,
            hard: PropTypes.number,
            deadly: PropTypes.number,
        }),
    }),
    name: PropTypes.string.isRequired,
    monster_ids: PropTypes.arrayOf( PropTypes.object ),
    challenge_rating: PropTypes.number,
    xp_rating: PropTypes.number,
    xp: PropTypes.number,
    challenge_modified: PropTypes.number,
    xp_modified: PropTypes.number,
};

TableRow.defaultProps = {
    id: null,
    campaign_id: null,
    hostedParty: null,
    name: '',
    monster_ids: [],
    challenge_rating: null,
    xp_rating: null,
    xp: null,
    challenge_modified: null,
    xp_modified: null,
};

export default ListDataWrapper(
    TableRow,
    ['hosted_party'],
    null,
    {hosted_party: 'hostedParty'}
);
