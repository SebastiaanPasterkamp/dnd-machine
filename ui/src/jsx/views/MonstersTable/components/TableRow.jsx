import React from 'react';
import PropTypes from 'prop-types';

import MonsterLabel from '../../../components/MonsterLabel';
import MonsterLinks from '../../../components/MonsterLinks';

const TableRow = function(props) {
    const { id } = props;
    return (
        <tr data-name={id}>
            <th>
                <MonsterLabel
                    monster={props}
                />
                <MonsterLinks
                    altStyle={true}
                    id={id}
                />
            </th>
            <td>
                <MonsterLabel
                    monster={props}
                    showName={false}
                    showType={true}
                />
            </td>
            <td>
                <MonsterLabel
                    monster={props}
                    showName={false}
                    showRating={true}
                />
            </td>
            <td>
                <MonsterLabel
                    monster={props}
                    showName={false}
                    showCampaign={true}
                />
            </td>
        </tr>
    );
};

TableRow.propTypes = {
    id: PropTypes.number.isRequired,
    campaign_id: PropTypes.number,
    name: PropTypes.string.isRequired,
    challenge_rating: PropTypes.number.isRequired,
    xp: PropTypes.number.isRequired,
};

TableRow.defaultProps = {
    campaign_id: null,
};

export default TableRow;
