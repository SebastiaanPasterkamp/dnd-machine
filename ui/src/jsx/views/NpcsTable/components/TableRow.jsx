import React from 'react';
import PropTypes from 'prop-types';

import NpcLinks from '../../../components/NpcLinks';
import CampaignLabel from '../../../components/CampaignLabel';

const TableRow = function({
    id, campaign_id, name, race, class: _class, alignment, location, organization,
}) {
    return (
        <tr data-name={id}>
            <th>
                {name}
                <NpcLinks
                    altStyle={true}
                    id={id}
                />

            </th>
            <td>{race}</td>
            <td>{_class}</td>
            <td>{alignment}</td>
            <td>{location}</td>
            <td>{organization}</td>
            <td>{campaign_id
                ? <CampaignLabel id={campaign_id} />
                : null
            }</td>
        </tr>
    );
};

TableRow.propTypes = {
    id: PropTypes.number.isRequired,
    campaign_id: PropTypes.number,
    name: PropTypes.string.isRequired,
    race: PropTypes.string,
    class: PropTypes.string,
    alignment: PropTypes.string,
    location: PropTypes.string,
    organization: PropTypes.string,
};

TableRow.defaultProps = {
    campaign_id: null,
};

export default TableRow;
