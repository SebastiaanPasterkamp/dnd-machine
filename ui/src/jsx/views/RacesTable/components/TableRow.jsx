import React from 'react';
import PropTypes from 'prop-types';
import MDReactComponent from 'markdown-react-js';

import RaceLinks from '../../../components/RaceLinks';

export const TableRow = function({ id, name, description }) {
    return (
        <tr data-name={id}>
            <th>
                { name }
                <RaceLinks
                    altStyle={true}
                    id={id}
                />
            </th>
            <td>
                <MDReactComponent
                    text={description || ''}
                />
            </td>
        </tr>
    );
};

TableRow.propTypes = {
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
    description: PropTypes.string,
};

TableRow.defaultProps = {
    description: '',
};

export default TableRow;
