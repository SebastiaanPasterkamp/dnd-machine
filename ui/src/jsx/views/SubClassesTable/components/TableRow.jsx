import React from 'react';
import PropTypes from 'prop-types';
import MDReactComponent from 'markdown-react-js';

import SubClassLinks from '../../../components/SubClassLinks';

export const TableRow = function({ id, name, description }) {
    return (
        <tr data-name={id}>
            <th>
                { name }
                <SubClassLinks
                    altStyle={true}
                    id={id}
                />
            </th>
            <td>
                <MDReactComponent
                    text={description || ''}
                />
            </td>
        </tr>
    );
};

TableRow.propTypes = {
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
    description: PropTypes.string,
};

TableRow.defaultProps = {
    description: '',
};

export default TableRow;
