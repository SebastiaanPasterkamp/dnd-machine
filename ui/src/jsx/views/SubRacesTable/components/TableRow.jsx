import React from 'react';
import PropTypes from 'prop-types';
import MDReactComponent from 'markdown-react-js';

import SubRaceLinks from '../../../components/SubRaceLinks';

export const TableRow = function({ id, name, description }) {
    return (
        <tr data-name={id}>
            <th>
                { name }
                <SubRaceLinks
                    altStyle={true}
                    id={id}
                />
            </th>
            <td>
                <MDReactComponent
                    text={description || ''}
                />
            </td>
        </tr>
    );
};

TableRow.propTypes = {
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
    description: PropTypes.string,
};

TableRow.defaultProps = {
    description: '',
};

export default TableRow;
