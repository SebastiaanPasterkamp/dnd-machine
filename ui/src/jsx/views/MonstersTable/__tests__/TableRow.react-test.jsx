import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import MockRouter from 'react-mock-router';

import ListDataActions from 'actions/ListDataActions';
import ObjectDataActions from 'actions/ObjectDataActions';
import {
    alignments,
    mockedApi,
    monster_types,
    size_hit_dice,
} from '../../../../../tests/__mocks__';

import TableRow from '../components/TableRow';

describe('TableRow', () => {
    const fullProps = {
        id: 1,
        name: "Bear",
        challenge_rating: 0.25,
        xp: 10,
    };

    beforeEach(() => {
        fetch.mockImplementation( mockedApi({
            alignments,
            campaign: {
                id: 2,
                name: "Circus",
                description: "Not your typical carnaval",
                user_id: 3,
            },
            current_user: {
                id: 1,
                role: ['dm'],
            },
            monster_types,
            size_hit_dice,
        }) );
        ListDataActions.fetchItems('alignments', 'items');
        ObjectDataActions.getObject('campaign', 2);
        ListDataActions.fetchItems('current_user');
        ListDataActions.fetchItems('monster_types', 'items');
        ListDataActions.fetchItems('size_hit_dice', 'items');
    })

    afterEach(() => fetch.resetMocks())

    it('should render', () => {
        const tree = renderer.create(
            <MockRouter>
                <TableRow {...fullProps} />
            </MockRouter>
        );

        expect(tree).toMatchSnapshot();
    });

    it('should render with campaign', () => {
        const tree = renderer.create(
            <MockRouter>
                <TableRow
                    {...fullProps}
                    campaign_id={2}
                />
            </MockRouter>
        );

        expect(tree).toMatchSnapshot();
    });

});
