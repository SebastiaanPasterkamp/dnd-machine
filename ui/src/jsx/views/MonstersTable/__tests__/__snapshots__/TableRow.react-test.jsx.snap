// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableRow should render 1`] = `
<tr
  data-name={1}
>
  <th>
    <div
      className="monster-label inline"
    >
      <span
        className="monster-label--name"
      >
        Bear
      </span>
    </div>
    <div
      className="base-link-group nice-btn-group"
    >
      <a
        className="cursor-pointer nice-btn-alt icon fa-eye"
        href={undefined}
        onClick={[Function]}
      >
        View
      </a>
      <a
        className="cursor-pointer nice-btn-alt icon fa-pencil"
        href={undefined}
        onClick={[Function]}
      >
        Edit
      </a>
      <a
        className="cursor-pointer nice-btn-alt icon fa-clone"
        onClick={[Function]}
      >
        Copy
      </a>
      <a
        className="cursor-pointer nice-btn-alt icon fa-trash-o bad"
        onClick={[Function]}
      >
        Delete
      </a>
    </div>
  </th>
  <td>
    <div
      className="monster-label inline"
    >
      <span
        className="monster-label--type"
      />
    </div>
  </td>
  <td>
    <div
      className="monster-label inline"
    >
      <span
        className="monster-label--rating"
      >
        <div
          className="challenge-rating inline"
        >
          CR 
          ¼
        </div>
         / 
        <div
          className="xp-rating inline"
        >
          10
           XP
        </div>
      </span>
    </div>
  </td>
  <td>
    <div
      className="monster-label inline"
    />
  </td>
</tr>
`;

exports[`TableRow should render with campaign 1`] = `
<tr
  data-name={1}
>
  <th>
    <div
      className="monster-label inline"
    >
      <span
        className="monster-label--name"
      >
        Bear
      </span>
    </div>
    <div
      className="base-link-group nice-btn-group"
    >
      <a
        className="cursor-pointer nice-btn-alt icon fa-eye"
        href={undefined}
        onClick={[Function]}
      >
        View
      </a>
      <a
        className="cursor-pointer nice-btn-alt icon fa-pencil"
        href={undefined}
        onClick={[Function]}
      >
        Edit
      </a>
      <a
        className="cursor-pointer nice-btn-alt icon fa-clone"
        onClick={[Function]}
      >
        Copy
      </a>
      <a
        className="cursor-pointer nice-btn-alt icon fa-trash-o bad"
        onClick={[Function]}
      >
        Delete
      </a>
    </div>
  </th>
  <td>
    <div
      className="monster-label inline"
    >
      <span
        className="monster-label--type"
      />
    </div>
  </td>
  <td>
    <div
      className="monster-label inline"
    >
      <span
        className="monster-label--rating"
      >
        <div
          className="challenge-rating inline"
        >
          CR 
          ¼
        </div>
         / 
        <div
          className="xp-rating inline"
        >
          10
           XP
        </div>
      </span>
    </div>
  </td>
  <td>
    <div
      className="monster-label inline"
    >
      <span
        className="monster-label--campaign"
      >
        <div
          className="campaign-label inline"
        >
          Circus
        </div>
      </span>
    </div>
  </td>
</tr>
`;
